devito/types/dense.py

```diff
@@ -103,7 +103,8 @@ def __eq__(self, other):
 
     _subs = Differentiable._subs
 
-    __hash__ = AbstractFunction.__hash__  # Required since we're overriding __eq__
+    def __hash__(self):
+        return id(self)
```

Comment:
ah, that's quite an interesting change

Reply:
I don't think the previous implementation is problematic in this case actually, since `__eq__` is tweaked in this class to define equality to be true iff the two objects are actually the same object. Which in turn means any hash function should work fine - but in that case `id` is the cheapest since it doesn't need to do anything. I was just on a bit of a crusade to make the `__hash__` implementations match the equality tests in case that was causing an issue.

I should put it next to the `__eq__` override, actually, as it's much clearer that it's correct when viewed next to that.

Or can get rid of it. I think I've realised since making the change that it's not important.